feat(routing-table): provide rt and routes functionality

h3adex opened the pull request:
## Description

This PR adds functionality to list and describe routing tables, as well as perform full CRUD operations on routes within those tables. It does not include support for creating or attaching routing tables, as those operations are currently intended to be managed exclusively through Terraform. 

This implementation is primarily aimed at enabling users to inspect and debug routes created via Terraform. Once the routing table feature reaches GA, support for creating routing tables and attaching them to networks will be added to the CLI.

<img width="794" height="447" alt="Screenshot 2025-10-10 at 16 22 40" src="https://github.com/user-attachments/assets/8ff57f7c-0410-47e0-bc02-c8422df9a658" />

## Checklist

- [x] Issue was linked above
- [x] Code format was applied: `make fmt`
- [x] Examples were added / adjusted (see e.g. [here](https://github.com/stackitcloud/stackit-cli/blob/ef291d1683ca5b0d719ec0a26ecb999a32685117/internal/cmd/ske/cluster/create/create.go#L49-L63))
- [x] Docs are up-to-date: `make generate-docs` (will be checked by CI)
- [x] Unit tests got implemented or updated
- [x] Unit tests are passing: `make test` (will be checked by CI)
- [x] No linter issues: `make lint` (will be checked by CI)

rubenhoenle commented on internal/cmd/beta/routingtable/list/list_test.go:
```diff
@@ -0,0 +1,215 @@
+var testOrgId = uuid.NewString()
+var testNetworkAreaId = uuid.NewString()
+
+var testLabelSelectorFlag = "key1=value1,key2=value2"
```
could be const

rubenhoenle commented on internal/cmd/beta/routingtable/list/list_test.go:
```diff
@@ -0,0 +1,215 @@
+
+	for _, tt := range tests {
+		t.Run(tt.description, func(t *testing.T) {
+			p := print.NewPrinter()
```
Please use the new `testutils.TestParseInput` func here to save us the boilerplate code.

https://github.com/stackitcloud/stackit-cli/blob/main/internal/cmd/image/list/list_test.go#L129

see also the PR https://github.com/stackitcloud/stackit-cli/pull/1033 for reference

rubenhoenle commented on internal/cmd/beta/routingtable/route/create/create_test.go:
```diff
@@ -0,0 +1,674 @@
+var testCtx = context.WithValue(context.Background(), testCtxKey{}, "foo")
+var testClient = &iaasalpha.APIClient{}
+
+var testRegion = "eu01"
```
could be const

rubenhoenle commented on internal/cmd/beta/routingtable/route/update/update_test.go:
```diff
@@ -0,0 +1,236 @@
+	"github.com/stackitcloud/stackit-sdk-go/services/iaasalpha"
+)
+
+var testRegion = "eu01"
```
could be const

rubenhoenle commented on internal/cmd/beta/routingtable/list/list.go:
```diff
@@ -0,0 +1,201 @@
+	return request
+}
+func outputResult(p *print.Printer, outputFormat string, items []iaasalpha.RoutingTable) error {
+	switch outputFormat {
```
Please use the new `Printer.OutputResult` func here to save us the boilerplate code for the JSON/YAML output.

https://github.com/stackitcloud/stackit-cli/blob/4725975b8aedb09a8fce802d7cdea63dafc187a2/internal/pkg/print/print.go#L247-L268

https://github.com/stackitcloud/stackit-cli/blob/4725975b8aedb09a8fce802d7cdea63dafc187a2/internal/cmd/image/list/list.go#L139-L178

see also the PR https://github.com/stackitcloud/stackit-cli/pull/1030 for reference

rubenhoenle commented on internal/cmd/beta/routingtable/list/list.go:
```diff
@@ -0,0 +1,201 @@
+				} else {
+					params.Printer.Debug(print.ErrorLevel, "configure resource manager client: %v", err)
+				}
+				params.Printer.Info("No routing-tables found for organization %q\n", orgLabel)
```
Invalid JSON/YAML output in case the `--output-format` flag is set to JSON/YAML output

rubenhoenle commented on internal/cmd/routingtable/list/list.go:
```diff
@@ -0,0 +1,198 @@
+	*globalflags.GlobalFlagModel
+	LabelSelector  *string
+	Limit          *int64
+	NetworkAreaId  *string
```
same here

rubenhoenle commented on internal/cmd/routingtable/create/create.go:
```diff
@@ -0,0 +1,208 @@
+				`stackit routing-table create --organization-id xxx --network-area-id yyy --name "rt" --description "some description"`,
+			),
+			examples.NewExample(
+				"Create a routing-table with name `rt` with system_routes disabled",
```
```suggestion
				"Create a routing-table with name `rt` with system routes disabled",
```

why the underscore?

rubenhoenle commented on internal/cmd/routingtable/list/list.go:
```diff
@@ -0,0 +1,198 @@
+				} else {
+					params.Printer.Debug(print.ErrorLevel, "configure resource manager client: %v", err)
+				}
+				params.Printer.Outputf("No routing-tables found for organization %q\n", orgLabel)
```
Invalid JSON/YAML output in case `-output-format` flag is set to JSON/YAML output

h3adex replied:
As a user, I still want to be able to list routing tables for automation purposes, such as listing all routing tables, grepping the IDs, and deleting them.

rubenhoenle replied:
Yeah, I also want users to be able to do that.

In fact running 

`stackit routing-table list --output-format json` currently would output `No routing-tables found for organization ...` when there are no routing tables present, but as a user I would expect a valid JSON output which would be just `[]` (an empty list).

rubenhoenle commented on internal/cmd/routingtable/route/describe/describe.go:
```diff
@@ -0,0 +1,195 @@
+const (
+	networkAreaIdFlag  = "network-area-id"
+	organizationIdFlag = "organization-id"
+	routeIdArg         = "ROUTE_ID_ARG"
```
```suggestion
	routeIdArg         = "ROUTE_ID"
```

rubenhoenle commented on internal/cmd/routingtable/update/update.go:
```diff
@@ -0,0 +1,167 @@
+	networkAreaIdFlag    = "network-area-id"
+	nonDynamicRoutesFlag = "non-dynamic-routes"
+	organizationIdFlag   = "organization-id"
+	routingTableIdArg    = "ROUTE_TABLE_ID_ARG"
```
```suggestion
	routingTableIdArg    = "ROUTING_TABLE_ID"
```

rubenhoenle replied:
Keep it consistent across your commands please

rubenhoenle commented on internal/cmd/routingtable/update/update.go:
```diff
@@ -0,0 +1,167 @@
+			}
+
+			// Call API
+			req := apiClient.UpdateRoutingTableOfArea(
```
Please create a `buildRequest` func which creates the request struct and adds the payload.

Then you can also unit-test this :wink: 

https://github.com/stackitcloud/stackit-cli/blob/98a09dd59c41467277dda114bcacc4084afb9fbc/internal/cmd/server/update/update.go#L118-L127

rubenhoenle commented on internal/cmd/routingtable/route/update/update.go:
```diff
@@ -0,0 +1,142 @@
+func parseInput(p *print.Printer, cmd *cobra.Command, inputArgs []string) (*inputModel, error) {
+	globalFlags := globalflags.Parse(p, cmd)
+
+	if len(inputArgs) == 0 {
```
Guess this should be handled already by the following line above

```golang
Args:  args.SingleArg(routeIdArg, utils.ValidateUUID),
```

rubenhoenle commented on internal/cmd/routingtable/list/list.go:
```diff
@@ -0,0 +1,198 @@
+
+	return p.OutputResult(outputFormat, items, func() error {
+		table := tables.NewTable()
+		table.SetHeader("ID", "NAME", "DESCRIPTION", "CREATED_AT", "UPDATED_AT", "DEFAULT", "LABELS", "SYSTEM_ROUTES", "DYNAMIC_ROUTES")
```
```suggestion
		table.SetHeader("ID", "NAME", "DESCRIPTION", "CREATED AT", "UPDATED AT", "DEFAULT", "LABELS", "SYSTEM ROUTES", "DYNAMIC ROUTES")
```

same here

rubenhoenle commented on internal/cmd/routingtable/list/list_test.go:
```diff
@@ -0,0 +1,179 @@
+		routingTable []iaas.RoutingTable
+		wantErr      bool
+	}{
+		{
```
Please add test cases with

```golang
// empty slice
routingTable: []iaas.RoutingTable{},

// empty routing table in slice (find nil pointer issues)
routingTable: []iaas.RoutingTable{{}},
```

rubenhoenle commented on internal/cmd/routingtable/route/list/list_test.go:
```diff
@@ -0,0 +1,240 @@
+
+	for _, tt := range tests {
+		t.Run(tt.description, func(t *testing.T) {
+			p := print.NewPrinter()
```
This is not resolved. Don't resolve comments before you've addressed them please. I'm having this problem with you in multiple of your PRs :disappointed:

rubenhoenle commented on internal/cmd/routingtable/route/describe/describe_test.go:
```diff
@@ -0,0 +1,209 @@
+		route        iaas.Route
+		wantErr      bool
+	}{
+		{
```
Same here, test the table output. JSON/YAML will pretty likely always work.

rubenhoenle commented on internal/cmd/routingtable/route/create/create_test.go:
```diff
@@ -0,0 +1,689 @@
+			items:        nil,
+			wantErr:      true,
+		},
+		{
```
Please add a test case with an empty route in the slice to find nil pointer dereferences

```golang
items:        []iaas.Route{{}},
```

rubenhoenle commented on internal/cmd/routingtable/route/create/create_test.go:
```diff
@@ -0,0 +1,689 @@
+	}{
+		{
+			name:         "nil items should return error",
+			outputFormat: "",
```
```suggestion
			outputFormat: print.PrettyOutputFormat,
```

rubenhoenle commented on internal/cmd/routingtable/route/create/create_test.go:
```diff
@@ -0,0 +1,689 @@
+		},
+		{
+			name:         "empty items list",
+			outputFormat: "",
```
```suggestion
			outputFormat: print.PrettyOutputFormat,
```

rubenhoenle commented on internal/cmd/routingtable/route/create/create_test.go:
```diff
@@ -0,0 +1,689 @@
+		},
+		{
+			name:         "table output with one route",
+			outputFormat: "",
```
```suggestion
			outputFormat: print.PrettyOutputFormat,
```

rubenhoenle commented on internal/cmd/routingtable/route/list/list_test.go:
```diff
@@ -0,0 +1,242 @@
+		routes       []iaas.Route
+		wantErr      bool
+	}{
+		{
```
same here, test your table output...

rubenhoenle commented on internal/cmd/routingtable/route/update/update_test.go:
```diff
@@ -0,0 +1,211 @@
+		outputFormat string
+		route        iaas.Route
+		wantErr      bool
+	}{
```
test your table output...

rubenhoenle replied:
These tests are kind of useless in their current state.

You must test edge cases, not only the happy paths for unit tests to be useful.

h3adex commented on internal/cmd/routingtable/delete/delete.go:
```diff
@@ -0,0 +1,111 @@
+const (
+	networkAreaIdFlag  = "network-area-id"
+	organizationIdFlag = "organization-id"
+	routingTableIdArg  = "ROUTING_TABLE_ARG"
```
I used it since it is stated in the contribution guide.

rubenhoenle commented on internal/cmd/routingtable/route/list/list.go:
```diff
@@ -0,0 +1,183 @@
+			}
+
+			if items := response.Items; items == nil || len(*items) == 0 {
+				params.Printer.Outputf("No routes  found for routing-table %q\n", model.RoutingTableId)
```
see other conversation

h3adex commented on internal/cmd/routingtable/create/create.go:
```diff
@@ -175,28 +167,18 @@ func outputResult(p *print.Printer, outputFormat string, routingTable *iaas.Rout
-
 		table := tables.NewTable()
-		table.SetHeader("ID", "NAME", "DESCRIPTION", "CREATED_AT", "UPDATED_AT", "DEFAULT", "LABELS", "SYSTEM_ROUTES", "DYNAMIC_ROUTES")
+		table.SetHeader("ID", "NAME", "DESCRIPTION", "DEFAULT", "LABELS", "SYSTEM ROUTES", "DYNAMIC ROUTES", "CREATED AT", "UPDATED AT")
```
I’ve reordered the table header. I thought it would be more readable to move `created_at` and `updated_at` to the end.

h3adex commented:
Rebased PR and added the option to update system_routes flag of a routing-table since it was added to the GO-SDK in the meanwhile: https://github.com/stackitcloud/stackit-sdk-go/blob/main/services/iaas/CHANGELOG.md#v130

rubenhoenle commented on internal/cmd/routingtable/route/delete/delete.go:
```diff
@@ -0,0 +1,113 @@
+				return err
+			}
+
+			if !model.AssumeYes {
```
```suggestion
			
```

shouldn't be needed anymore after the integration of #1217

rubenhoenle commented on internal/cmd/routingtable/create/create.go:
```diff
@@ -0,0 +1,190 @@
+	nameFlag             = "name"
+	networkAreaIdFlag    = "network-area-id"
+	nonDynamicRoutesFlag = "non-dynamic-routes"
+	nonSystemRoutesFlag  = "non-system-routes"
```
same here

rubenhoenle commented on internal/cmd/routingtable/update/update.go:
```diff
@@ -0,0 +1,174 @@
+	labelFlag            = "labels"
+	nameFlag             = "name"
+	networkAreaIdFlag    = "network-area-id"
+	nonDynamicRoutesFlag = "non-dynamic-routes"
```
```suggestion
	dynamicRoutesFlag = "dynamic-routes"
```

Same here, I don't get the `non-` thing.

rubenhoenle commented:
@h3adex Could you please resolve the open comments?
